libcontainer/cgroups/file.go

```diff
@@ -109,43 +114,43 @@ func prepareOpenat2() error {
 	return prepErr
 }
 
-// OpenFile opens a cgroup file in a given dir with given flags.
-// It is supposed to be used for cgroup files only.
 func openFile(dir, file string, flags int) (*os.File, error) {
 	mode := os.FileMode(0)
 	if TestMode && flags&os.O_WRONLY != 0 {
 		// "emulate" cgroup fs for unit tests
 		flags |= os.O_TRUNC | os.O_CREATE
 		mode = 0o600
 	}
+	path := path.Join(dir, file)
 	if prepareOpenat2() != nil {
-		return openFallback(dir, file, flags, mode)
+		return openFallback(path, flags, mode)
 	}
-	reldir := strings.TrimPrefix(dir, cgroupfsPrefix)
-	if len(reldir) == len(dir) { // non-standard path, old system?
-		return openFallback(dir, file, flags, mode)
+	relPath := strings.TrimPrefix(path, cgroupfsPrefix)
+	if len(relPath) == len(path) { // non-standard path, old system?
+		return openFallback(path, flags, mode)
```

Comment:
To help reviewers.

Basically, the code previously assumed that `dir` starts with `/sys/fs/cgroup/` (including the ending slash), which is not always the case (such as when calling `OpenFile("/sys/fs/cgroup", "cgroup.controllers")`). So, the fallback was used for this case, which is not correct.

This commit removes such assumption,  using `path.Join(dir, file)` as a full path. This simplifies the code, and helps the case above, as well as something like `OpenFile("/sys", "fs/cgroup/cgroup.controllers")`.